haskell/test/SMT/BaseDAO/Permit.hs

```diff
@@ -0,0 +1,42 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+module SMT.BaseDAO.Permit
+ ( verifyPermitProtectedVote
+ ) where
+
+import Universum
+
+import Lorentz hiding (not, cast, checkSignature)
+import Tezos.Address (mkKeyAddress)
+import Tezos.Crypto (checkSignature)
+
+import Ligo.BaseDAO.Types
+import SMT.Types
+
+voteParamToSignedData :: ContractCall Parameter -> VoteParam -> Storage -> (ByteString, Storage)
+voteParamToSignedData cc param store =
+  ( lPackValueRaw @((ChainId, Address), (Nonce, VoteParam))
+    ( (cc & ccChainId, cc & ccSelfAddress)
+    , (store & sPermitsCounter, param)
+    )
+  , store { sPermitsCounter = Nonce (unNonce (store & sPermitsCounter) + 1) }
+  )
+
+checkedPermitSender :: Permit VoteParam -> ByteString -> Either ModelError Address
+checkedPermitSender permit dataToSign =
+  if checkSignature (permit & pKey) (permit & pSignature & unTSignature) dataToSign then
+    Right (mkKeyAddress (permit & pKey))
+  else Left MISSIGNED
+
+verifyPermitVote :: ContractCall Parameter -> Permit VoteParam -> VoteParam -> Storage -> Either ModelError (VoteParam, Address, Storage)
+verifyPermitVote cc permit voteParam store = do
+  let (dataToSign, updatedStore) = voteParamToSignedData cc voteParam store
+  permitSender <- checkedPermitSender permit dataToSign
+  Right (voteParam, permitSender, updatedStore)
+
+verifyPermitProtectedVote :: ContractCall Parameter -> PermitProtected VoteParam -> Storage -> Either ModelError (VoteParam, Address, Storage)
+verifyPermitProtectedVote cc permited store = do
+  case (permited & ppPermit) of
+    Nothing -> Right (permited & ppArgument, cc & ccSender, store)
+    Just permit -> verifyPermitVote cc permit (permited & ppArgument) store
```

Comment:
I include it simply to make sure permits work. I think even though it does not take advantage of the generator, it is still worth including it. Unless it is necessary to remove it, shouldn't we keep it?

Reply:
I also think we should definitely incluse the generation and check of permits.

Reply:
>incluse the generation and check of permits.

It appears to me that it would just be redundant because we already have tests that check this. And IIUC, if those checks pass, any sequence of calls that include permit wrapped calls in them will work exactly the same without those calls being wrapped in permit.

Reply:
So I've added the `vote` entrypoint call that include randomize of permit. It has the chances of being: no permit, correct permit and wrong permit. Hopefully, this is good enough, and won't affect the effectiveness of the tests. You can take a look at this.